java/server/src/org/openqa/selenium/grid/sessionqueue/GetNewSessionResponse.java

```diff
@@ -0,0 +1,135 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.selenium.grid.sessionqueue;
+
+import static java.net.HttpURLConnection.HTTP_INTERNAL_ERROR;
+import static org.openqa.selenium.grid.data.NewSessionRejectedEvent.NEW_SESSION_REJECTED;
+import static org.openqa.selenium.grid.data.NewSessionResponseEvent.NEW_SESSION_RESPONSE;
+import static org.openqa.selenium.remote.http.Contents.asJson;
+import static org.openqa.selenium.remote.http.Contents.bytes;
+
+import com.google.common.collect.ImmutableMap;
+
+import org.openqa.selenium.events.EventBus;
+import org.openqa.selenium.grid.data.NewSessionErrorResponse;
+import org.openqa.selenium.grid.data.NewSessionRequest;
+import org.openqa.selenium.grid.data.NewSessionResponse;
+import org.openqa.selenium.internal.Require;
+import org.openqa.selenium.remote.http.HttpRequest;
+import org.openqa.selenium.remote.http.HttpResponse;
+import org.openqa.selenium.remote.tracing.Tracer;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class GetNewSessionResponse {
+
+  private static final Logger LOG = Logger.getLogger(GetNewSessionResponse.class.getName());
+  private final EventBus bus;
+  private final Tracer tracer;
+  public final NewSessionQueue sessionRequests;
+  private final Map<UUID, NewSessionRequest> knownRequests = new ConcurrentHashMap<>();
+
+  public GetNewSessionResponse(Tracer tracer, EventBus bus,
+                               NewSessionQueue sessionRequests) {
+    this.tracer = Require.nonNull("Tracer", tracer);
+    this.bus = Require.nonNull("Event bus", bus);
+    this.sessionRequests = Require.nonNull("New Session Request Queue", sessionRequests);
+
+    this.bus.addListener(NEW_SESSION_RESPONSE, event -> {
+      try {
+        NewSessionResponse sessionResponse = event.getData(NewSessionResponse.class);
+        this.setResponse(sessionResponse);
+      } catch (Exception ignore) {
+        // Ignore any exception. Do not want to block the eventbus thread.
+      }
+    });
+
+    this.bus.addListener(NEW_SESSION_REJECTED, event -> {
+      try {
+        NewSessionErrorResponse sessionResponse = event.getData(NewSessionErrorResponse.class);
+        this.setErrorResponse(sessionResponse);
+      } catch (Exception ignore) {
+        // Ignore any exception. Do not want to block the eventbus thread.
+      }
+    });
+  }
+
+  private void setResponse(NewSessionResponse sessionResponse) {
+    // Each thread will get its own CountDownLatch and it is stored in the Map using request id as the key.
+    // EventBus thread will retrieve the same request and set it's response and unblock waiting request thread.
+    UUID id = sessionResponse.getRequestId();
+    NewSessionRequest sessionRequest = knownRequests.get(id);
```

Comment:
If `remove` is called before `setResponse` this will cause an NPE.

Reply:
`remove` will be called only after the CountdownLatch countdowns, which happens only when a response is received.